File: src/Go/killer.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"sort"
	"time"
)

func main() {
	
	start := time.Now()
	// Open the file
	csvfile, err := os.Open("/home/sandy/Documents/puzzle.csv")
	if err != nil {
		log.Fatalln("Couldn't open the csv file", err)
	}

	// Parse the file
	r := csv.NewReader(csvfile)
	//r := csv.NewReader(bufio.NewReader(csvfile))
	r.FieldsPerRecord = -1
	
	
	var parms [][]int
	
	// Iterate through the records
	for {
		// Read each record from csv
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		var recordInt []int
		for _,j := range record {
		   tempInt, err := strconv.Atoi(j)
		   if err == nil {
		   recordInt = append(recordInt,tempInt)
			}
		}
		parms = append(parms,recordInt)
		//fmt.Println(record)
	}
	sort.SliceStable(parms, func(i, j int) bool {
		return parms[i][len(parms[i])-1] < parms[j][len(parms[j])-1]
	})

	fmt.Println(parms)

	var startBoard [81]int


	recursiveCheck(startBoard, parms)
	
	end := time.Now()
	fmt.Println(start)
	fmt.Println(end)
}

func FilterParms(vs [][]int, f func([]int) bool) []int {
	vsf := make([]int, 0)
	for _, v := range vs {
		if f(v) {
			vsf = v
			return vsf
		}
	}
	return vsf
}
func sameRow(i int, j int) bool {
	return (i/9)>>0 == (j/9)>>0
}
func sameCol(k int, l int) bool {
	return (k-l)%9 == 0
}
func sameBlock(m int, n int) bool {
	return ((m/27)>>0 == (n/27)>>0 && ((m%9)/3)>>0 == ((n%9)/3)>>0)
}
func IndexOf(array [81]int, val int) int {
	for i, v := range array {
		if v == val {
			return i
		}
	}
	return -1
}

func checkRow(aRow [81]int, parms [][]int) bool {
	var pos = IndexOf(aRow, 0)
	var selectedParm = (FilterParms(parms, func(v []int) bool {
		return v[len(v)-1] == pos-1
	}))
//	fmt.Println(selectedParm)

	if len(selectedParm) > 0 {
		var total = 0
		for i, j := range selectedParm {
			if i > 0 {
				total = total + aRow[j]
			}
		}
//		fmt.Println(total)
		if total != selectedParm[0] {
			return false
		} else {
			return true
		}
	} else {
		return true
	}
}
func recursiveCheck(inBoard [81]int, parms [][]int) bool {
	var i = IndexOf(inBoard, 0)
	if i == -1 {
		fmt.Println(inBoard)
		return true
	}

	var excludedNumbers = make(map[int]bool)
	var x int = 0
	var y int = 1

	for x < i {
		if sameRow(i, x) || sameCol(i, x) || sameBlock(i, x) {
			excludedNumbers[inBoard[x]] = true
		}

		x++
	}

	for y < 10 {
		if !excludedNumbers[y] {
			var newBoard [81]int 
			copy( newBoard[:],inBoard[0:81])
			newBoard[i] = y
			if checkRow(newBoard,parms) {
			recursiveCheck(newBoard, parms)
			}
		}
		y++
	}
	return true
}

File: src/lua/Killer.lua
function index(array,value)
	for i=0,#array do
		if array[i] == value then
			return i
		end	
	end
end

function table.copy(input)
	local output = {}
	for i=0,#input do
		output[i] = input[i]
	end
return output
end		

function get_rules()
	local rules = {}    
	local f = io.open("puzzle.csv","r")
	for line in f:lines() do
		local temp = {}
		for word in line:gmatch('[^,%s]+') do
			table.insert(temp, tonumber(word))
		end	
		rules[tonumber(temp[#temp])] = temp
	end
	return rules
end

function get_startBoard()
	local startBoard = {}
    for i= 0, 80 do
      startBoard[i] = 0
    end
    return startBoard
end    

function same_row(i,j)
    return (i // 9)  == (j // 9) ;
end

function same_col(k,l)
	if k > l then return (k - l) % 9 == 0
	else return (l - k) % 9 == 0
	end
end

function same_block(m,n)
        return (m // 27) == (n // 27) and ((m % 9) // 3) == ((n % 9) // 3) 
end

function get_validCombo()    
validCombo = {}          
    for i=0,80 do
      validCombo[i] = {}     
      for j=0,80 do
        validCombo[i][j] = same_row(i,j) or same_col(i,j) or same_block(i,j)
      end
    end
    return validCombo
end       

function checkRow(aRow, position, value) 

	local selectedRule = rules[position]
	if selectedRule ~= nil then
		for i,val in ipairs(selectedRule) do
				if i > 1 then
					value = value + aRow[val]
				end	
			end	
			if value ~= selectedRule[1] then
				return false
			else
				return true
			end
	return true
	end
return true
end

function recursiveCheck(inBoard)
	
	local pos = index(inBoard, 0)
	if pos == nil then
		print("finished")
	else	
		local excludedNumbers = {}
		local x = 0
		local y = 1
		while x < pos do
			if validCombos[pos][x] then 
				excludedNumbers[inBoard[x]] = true
			end	
			x = x + 1
		end

		while y < 10 do
			if excludedNumbers[y] == nil then	
				if pos == 80 then
					inBoard[80] = y
					local str = ""
					for i=0,#inBoard do
						str = str.." ".. inBoard[i]
					end
					print(str)	
					print(string.format("elapsed time: %.2f\n", os.clock() - x))
					os.exit()
				end	
				if checkRow(inBoard,pos,y) then
				    local newBoard = table.copy(inBoard)
					newBoard[pos] = y 
					recursiveCheck(newBoard)
				end	
			end
		y = y + 1
		end
	end
	return true
end


x = os.clock()
rules = get_rules()
validCombos = get_validCombo()
local startBoard = get_startBoard()
recursiveCheck(startBoard)


